uniffi_core/src/ffi/rustfuture/scheduler.rs

```diff
@@ -0,0 +1,94 @@
+///   state, invoking any future callbacks as soon as they're stored.
+
+#[derive(Debug)]
+pub(super) enum Scheduler {
+    /// No continuations set, neither wake() nor cancel() called.
+    Empty,
+    /// `wake()` was called when there was no continuation set.  The next time `store` is called,
+    /// the continuation should be immediately invoked with `RustFuturePoll::MaybeReady`
+    Waked,
+    /// The future has been cancelled, any future `store` calls should immediately result in the
+    /// continuation being called with `RustFuturePoll::Ready`.
+    Cancelled,
+    /// Continuation set, the next time `wake()`  is called is called, we should invoke it.
+    Set(RustFutureContinuationCallback, *const ()),
+}
+
+impl Scheduler {
+    pub(super) fn new() -> Self {
+        Self::Empty
+    }
+
+    /// Store new continuation data if we are in the `Empty` state.  If we are in the `Waked` or
+    /// `Cancelled` state, call the continuation immediately with the data.
+    pub(super) fn store(&mut self, callback: RustFutureContinuationCallback, data: *const ()) {
+        match self {
+            Self::Empty => *self = Self::Set(callback, data),
+            Self::Set(old_callback, old_data) => {
+                log::error!(
+                    "store: observed `Self::Set` state.  Is poll() being called from multiple threads at once?"
+                );
+                old_callback(*old_data, RustFuturePoll::Ready);
+                *self = Self::Set(callback, data);
+            }
+            Self::Waked => {
+                *self = Self::Empty;
+                callback(data, RustFuturePoll::MaybeReady);
+            }
+            Self::Cancelled => {
+                callback(data, RustFuturePoll::Ready);
+            }
+        }
+    }
+
+    pub(super) fn wake(&mut self) {
+        match self {
+            // If we had a continuation set, then call it and transition to the `Empty` state.
+            Self::Set(callback, old_data) => {
+                let old_data = *old_data;
+                let callback = *callback;
+                *self = Self::Empty;
+                callback(old_data, RustFuturePoll::MaybeReady);
+            }
+            // If we were in the `Empty` state, then transition to `Waked`.  The next time `store`
+            // is called, we will immediately call the continuation.
+            Self::Empty => *self = Self::Waked,
+            // This is a no-op if we were in the `Cancelled` or `Waked` state.
+            _ => (),
+        }
+    }
+
+    pub(super) fn cancel(&mut self) {
+        if let Self::Set(callback, old_data) = mem::replace(self, Self::Cancelled) {
+            callback(old_data, RustFuturePoll::Ready);
+        }
+    }
+
+    pub(super) fn is_cancelled(&self) -> bool {
+        matches!(self, Self::Cancelled)
+    }
+}
+
+// ContinuationDataCell is Send + Sync as long we handle the *const () pointer correctly
+
+unsafe impl Send for Scheduler {}
+unsafe impl Sync for Scheduler {}
```

Comment:
I think at least for `Send` the comment is wrong (and `Sync` would be trivially obtained by wrapping the `*const ()` in `sync_wrapper::SyncWrapper` which would not restrict any of the uses here).

For `Send` to hold for regular `&reference`s, the type being referenced needs to be `Sync` so I think this also needs to hold for any data we store in this pointer from any backend, and that should be documented (here, but also in any docs we have for backend authors).

Reply:
Good point.  I just updated that comment.  I'm also going to do a general rework on handles in another PR, I'll make sure to mention something like this there.  In general, I think whenever either side passes a handle over the FFI, it must be `Send` and the handled object must by `Sync`.